src/Pim/Component/Catalog/Completeness/CompletenessCalculator.php

```diff
@@ -0,0 +1,205 @@
+use Pim\Component\Catalog\Model\Completeness;
+use Pim\Component\Catalog\Model\CompletenessInterface;
+use Pim\Component\Catalog\Model\FamilyInterface;
+use Pim\Component\Catalog\Model\ProductInterface;
+use Pim\Component\Catalog\Model\ProductValueCollection;
+use Pim\Component\Catalog\Model\ProductValueCollectionInterface;
+use Pim\Component\Catalog\Model\ProductValueInterface;
+
+/**
+ * @author    Damien Carcel (damien.carcel@akeneo.com)
+ * @copyright 2017 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php Open Software License (OSL 3.0)
+ */
+class CompletenessCalculator implements CompletenessCalculatorInterface
+{
+    /** @var ProductValueFactory */
+    protected $productValueFactory;
+
+    /** @var CachedObjectRepositoryInterface */
+    protected $channelRepository;
+
+    /** @var CachedObjectRepositoryInterface */
+    protected $localeRepository;
+
+    /** @var ProductValueCompleteCheckerInterface */
+    protected $productValueCompleteChecker;
+
+    /**
+     * @param ProductValueFactory                  $productValueFactory
+     * @param CachedObjectRepositoryInterface      $channelRepository
+     * @param CachedObjectRepositoryInterface      $localeRepository
+     * @param ProductValueCompleteCheckerInterface $productValueCompleteChecker
+     */
+    public function __construct(
+        ProductValueFactory $productValueFactory,
+        CachedObjectRepositoryInterface $channelRepository,
+        CachedObjectRepositoryInterface $localeRepository,
+        ProductValueCompleteCheckerInterface $productValueCompleteChecker
+    ) {
+        $this->productValueFactory = $productValueFactory;
+        $this->channelRepository = $channelRepository;
+        $this->localeRepository = $localeRepository;
+        $this->productValueCompleteChecker = $productValueCompleteChecker;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function calculate(ProductInterface $product)
+    {
+        if (null === $product->getFamily()) {
+            return [];
+        }
+
+        $completenesses = [];
+        $requiredProductValueCollectionsList = $this->getRequiredProductValueCollections($product->getFamily());
+        $actualValues = $product->getValues();
+
+        foreach ($requiredProductValueCollectionsList as $channelCode => $requiredProductValueCollections) {
+            foreach ($requiredProductValueCollections as $localeCode => $requiredProductValueCollection) {
+                $completenesses[$channelCode][$localeCode] = $this->generateCompleteness(
+                    $product,
+                    $requiredProductValueCollection,
+                    $actualValues,
+                    $channelCode,
+                    $localeCode
+                );
+            }
+        }
+
+        return $completenesses;
+    }
+
+    /**
+     * Generates a two dimensional array indexed by scope and locale containing the required product value collections.
```

Comment:
explained that required values are determined from the attribute requirements of the family and from the channel